[ADR-5] - Subscription Module (#4780)
cosmos/cosmos-sdk: docs/architecture/adr-005-subscription.md

# ADR 5: Subscription

## Changelog

## Context

Cosmos-SDK should allow developers to choose from a variety of payment models for their application. One of the most popular payment models for Internet applications, so it is natural to provide this payment model as a module for sdk users.

For fuller context around this issue: see [\#4642](https://github.com/cosmos/cosmos-sdk/issues/4642)

## Decision

Create Subscription module in SDK so that users can subscribe to various subscription services. Subscription collectors will send messages to collect subscription payments from users who have due subscriptions. Subscriptions are paid out directly from a user account. If the user does not have enough funds to pay for the subscription is inactivated. Users can define maximum limits on how many periods the subscription is valid for; if this limit elapses before a subscription renewal, the subscription is invalidated.

## Status

#### Proposed

## Implementation Changes

Introduce the following messages to `x/subscription` module:

```go
// CreateSubscriptionMsg allows Service Provider to create new subscription service.

> type CreateSubscriptionMsg struct {

Author: Haven't thought about how to allow collector to update terms. Should it be enforced as soon as the message is sent (applying to all current subscriptions in duequeue) or should it apply to all subscriptions added to due queue after the message has been sent?

>     Name      string         // unique, human-readable name for subscription service

Reviewer: I think this shouldn't be unique and we should have an independent unique `string ID`. The `Name` would be a separate property that's not necessarily unique

    Amount    sdk.Coins      // amount to be collected for each subscription period

>     Period    time.time      // duration of subscription period

Reviewer: this should be `time.Duration`. 

Also, most of the subscriptions allow for a set of subscription periods (1mo, 3mo, 6mo, 1y) with their corresponding prices (i.e registering an `Amount` to a `Period`). 

I think it'd be better to extend this to be an array. Otherwise, the user will have to create multiple messages for each of these periods.

Author: Hmm yea, so the current FIFO queue currently works with the assumption that there is a single period and the Collector never gets to change (at least never shorten) this period.

Probably too restrictive. Would like to hear more about the tree idea, a bit fuzzy on what you mean.

    Collector sdk.AccAddress // address that will collect subscription payments
}
```

```go
// SubscribeMsg allows subscriber account to subscribe to provided service
// If subscription to the service from user already exists, this msg is treated as renewal.
// If msg.Limit != -1, then subscription.Limit += msg.Limit. Else, subscription.Limit = -1 (unlimited)
type SubscribeMsg struct {

>     Name       string         // name of service to subscribe to

Reviewer: replace with `SubscriptionID`. See comment above

    Subscriber sdk.AccAddress // address of subscriber

>     Limit      int64          // Maximum number of periods that subscription remains active. Limit = -1 implies no limit

Author: Should I interpret `-1` as unlimited or use a separate boolean field?

Reviewer: I think this is ok as is. What'd be missing is to choose a valid period of the subscription:
```go
type SubscribeMsg struct {
    Name       string         // name of service to subscribe to
    Subscriber sdk.AccAddress // address of subscriber
    Limit      int64          // Maximum number of periods that subscription remains active. Limit = -1 implies no limit
    Period     time.Duration  // chosen period after which the subscription amount is collected
}
```

>     Limit      int64          // Maximum number of periods that subscription remains active. Limit = -1 implies no limit

Reviewer: also, do we delete the subscription from the store once the limit is due? or do we move them to some other queue or smth and then we prune? 

I believe this requires a comment

Author: I was currently thinking just inactivate. If you look at the Collect handler, we pop the subscription off the queue if it is due (basically deleting from store). If the subscription is paid, it gets added back to end of queue.

If it is inactivated, the subscriptionID doesn't go back to end of queue, but it does stay in store as inactivated item.

Author: Ok so the reason I didn't delete is because I didn't want to traverse through the `DueQueue` and remove the deleted element. 

Though I think I can leave the subscriptionID in and just check if the retrieved subscription is nil. If it is, then I can remove it from the list

}
```

```go
// UnsubscribeMsg allows subscriber to inactivate an active subscription
type UnsubscribeMsg struct {

>     Name       string         // name of service to unsubscribe to

Reviewer: ditto

    Subscriber sdk.AccAddress // address of subscriber
}
```

```go
// CollectMsg allows a Collector for a service to collect payments on due subscriptions that are processed off a FIFO queue.
type CollectMsg struct {
    Name      string         // name of subscription service to collect payments from
    Collector sdk.AccAddress // address that will collect payments

>     Limit     int64          // maximum number of items to process in FIFO duequeue. If Limit = -1, try to process all due subscriptions

Author: Ditto here

}
```

Create a new store with the following key-values:

`Address => []SubscriptionID // List of active/inactive subscriptions owned by the users`

`Metadata:{Name} => SubscriptionMetaData`

> `DueQueue:{Name} => []SubscriptionID // FIFO queue of subscriptions. CONTRACT: All due subscriptions exist before all undue subscriptions. All subscriptions in DueQueue are active`

Reviewer: Idea for single `Period` only: 

As the tree is deterministic and we don't need to keep track of two lists, I think an alternative solution would be deleting the subscription from the store and then adding it again. That'd give us a list of subscriptions sorted by due time.

Author: The second list for `Address => []SubscriptionID` is so that users can easily query fo all subscriptions both active and inactive that their account has subscribed to. Doesn't have anything to do with collection, it is for help with querying

`SubscriptionID => Subscription`

```go
// SubscriptionID is a unique identifier to a subscription struct
// hash(Name+Address)
type SubscriptionID []byte
```

```go
// Subscription Metadata contains information necessary for processing subscriptions

> type Metadata struct {

Reviewer: I'd maybe rename this to `Terms`

Reviewer: ditto on missing `Description` and `Name` field

    Amount    sdk.Coins      // amount to be collected for each subscription period
    Period    time.time      // duration of subscription period
    Collector sdk.AccAddress // address that will collect subscription payments
}
```

```go
// Represents a user's subscription to a service
type Subscription {

>     Name       string

Author: Is it fair to enforce a unique name or should I assign a unique ID for each subscription and then allow for repeated names

Reviewer: unique hash ID

    Subscriber sdk.AccAddress
    Limit      int64
    LastPaid   time.Time

>     Active     bool

Reviewer: see the comment above. I think we can delete this field if we delete the subscription from the store once the limit was reached

Author: I'd like users to be able to query for subscriptions that have went inactive in the store and then choose to renew whatever inactive subscriptions they wish

Reviewer: they can still query the subscription `Terms`

Reviewer: + you can still query the events generated by the transactions

        return err
    }

    // check if user already has subscribe to this service
    // if so renew the subscription
    if subscription := GetUserSubscriptions(msg.Subscriber, msg.Name); subscription != nil {
        if msg.Limit == -1 {
            subscription.Limit = -1
        } else {
            subscription.Limit += msg.Limit
        }
        // set active to true
        subscription.Active = true
        StoreSubscription(subscriptionID, subscription)
        return nil
    }   
    subscription := Subscription{
        Name:       metadata.Name,
        Subscriber: msg.Subscriber,
        Limit:      msg.Limit,
        LastPaid:   ctx.BlockTime,
        Active:     true
    }
    StoreSubscription(subscriptionID, subscription)
    // adds subscription to list of user subscriptions
    AppendToUserSubscriptions(subscriptionID)
    // push to back of DueQueue for this service
    PushToBackOfDueQueue(msg.Name, subscriptionID)
}
```

```go
func HandleUnsubscribeMsg(ctx sdk.Context, msg UnsubscribeMsg) {
    subscription := GetUserSubscriptions(msg.Subscriber, msg.Name)
    if subscription == nil {
        return nil
    }
    subscription.Active = false

>     subscription.Limit = 0 // to ensure that renewal updates limit correctly

Author: Should I be resetting Limit so that any future Renewal will define the new limit in the message?

    subscriptionID := hash(msg.Name|msg.Subscriber)
    StoreSubscription(subscriptionID, subscription)
}
```

```go
func HandleCollectMsg(ctx sdk.Context, msg CollectMsg) {
    if !SubscriptionExists(msg.Name) {
        return NoSuchSubscriptionErr
    }
    metadata := GetMetaData(ctx, msg.Name)
    if msg.Collector != metadata.Collector {
        return InvalidCollectorErr
    }
    queue := GetDueQueue(msg.Name)
    for i := 0; i < msg.Limit; i++ {
        // top of the queue is "due"
        subscription := GetSubscription(queue[0])
        if subscription.LastPaid + metadata.Period >= ctx.BlockTime {

>             subscriptionID := PopOffFront(queue)

Author: I believe this can be made O(1). the index here was just for describing the idea. I think a very, very simple linked list will do just fine.

Each subscriptionID can be embedded in a `QueueItem` struct that has a pointer to the next element. The Queue itself just maintains a pointer to the first and last element of the queue. This would make this code O(1) since we're just updating pointers at front and end of queue.

By tree do you mean make use of the underlying IAVL tree and use `IteratePrefix` somehow?
What would be the key that each subscription gets stored under? So deleting and reinserting into tree will make insertion order into tree be the same as the processing order, but im not 100% sure if iterating prefix will give back the leaves in the insertion order they were submitted in.

>             subscriptionID := PopOffFront(queue)

Reviewer: I like the idea of the linked list. I think it's worth it to write it down in this document

>             subscriptionID := PopOffFront(queue)

Author: Hmm but if it's different Periods, this is no longer as simple to understand since we can't simply append to end of linked list. We might have to reverse-iterate a few items before placing it in the correct place.

What is the harm in maintaining a Queue for each period-length? Reduces the complexity, easier to analyze runtime and not too much overload i think since it's taking the same number of elements and splitting them across a small number of lists (maybe 3 or 4) rather than one big one.

>             subscriptionID := PopOffFront(queue)

Reviewer 2: I like the idea of maintaining a linked list queue for each period. Keeping collection at O(1) will be a nice optimization.

            if subscription.Limit == 0 {
                subscription.Active = false
                StoreSubscription(subscriptionID, subscription)
                continue
            }
            err := bank.Send(subscription.Subscriber, metadata.Collector, metadata.Amount)
            if err != nil {
                // insufficient funds. inactivate subscription
                subscription.Active = false

>                 subscription.Limit = 0

Author: Ditto here. Should I be resetting to 0

            } else {
                // update subscription after successful payment
                subscription.LastPaid = ctx.BlockTime
                if subscription.Limit != -1 {
                    subscription.Limit--
                }
            }
            // store updated subscription
            StoreSubscription(subscriptionID, subscription)
            // push paid subscription back to end of queue

>             PushToBack(queue, subscriptionID)

Reviewer: this _only_ works if all the subscriptions have the same `Period`. Otherwise, you need to sort them again

Author: This is correct

>             PushToBack(queue, subscriptionID)

Reviewer 2: No need to readd subscription if payment failed. Maybe move to inside else statement?

Author: it doesn't get added back. If the payment failed, i have a continue statement that skips the rest of the for loop body

Reviewer 2: yup, this issue got fixed in the commit below

# ADR 5: Subscription

## Changelog

## Context

> Cosmos-SDK should allow developers to choose from a variety of payment models for their application. Subscriptions areo ne of the most popular payment models for Internet applications, so it is natural to provide this payment model as a module for sdk users.

Reviewer 3: ```suggestion
Cosmos-SDK should allow developers to choose from a variety of payment models for their application. Subscriptions are one of the most popular payment models for Internet applications, so it is natural to provide this payment model as a module for sdk users.
```

For fuller context around this issue: see [\#4642](https://github.com/cosmos/cosmos-sdk/issues/4642)

## Decision

> Create Subscription module in SDK so that users can subscribe to various subscription services. Subscription collectors will send messages to collect subscription payments from users who have due subscriptions. Subscriptions are paid out directly from a user account. If the user does not have enough funds to pay for the subscription is inactivated. Users can define maximum limits on how many periods the subscription is valid for; if this limit elapses before a subscription renewal, the subscription is invalidated.

Reviewer 3: Very important to give the power of cancelling subscriptions to the subscribing users not the receiver of subscription funds.. maybe worth adding this design consideration explicitly

Author: The only way a receiver of subscription funds can "cancel" a subscription is if the subscribing account has insufficient funds. This seems reasonable to me as the user can always refund the account and resubscribe.

The maximum limit is a user-defined parameter so that they can define an expiration date for their subscription. This is to mitigate the common problem of unused subscriptions siphoning off funds from account without user noticing.

        return nil
    }

    // new subscribeMsg will pay for the first period
    err := bank.Send(msg.Subscriber, Terms.Collector, Terms.Amount)
    if err != nil {
        return err
    }

    subscription := Subscription{
        Name:       Terms.Name,
        Subscriber: msg.Subscriber,
        Limit:      msg.Limit,
        LastPaid:   ctx.BlockTime,
    }
    StoreSubscription(subscriptionID, subscription)
    // adds subscription to list of user subscriptions
    AppendToUserSubscriptions(subscriptionID)
    // push to back of DueQueue for this service
    PushToBackOfDueQueue(msg.Name, subscriptionID)
}
```

```go
// simply deletes mapping subscriptionID => subscription and removes ID from user list
// will not mutate duequeue
func HandleUnsubscribeMsg(ctx sdk.Context, msg UnsubscribeMsg) {
    DeleteSubscription(msg.Subscriber, msg.Name)
}
```

```go
func HandleCollectMsg(ctx sdk.Context, msg CollectMsg) {
    if !SubscriptionExists(msg.Name) {
        return NoSuchSubscriptionErr
    }
    Terms := GetTerms(ctx, msg.Name)
    if msg.Collector != Terms.Collector {
        return InvalidCollectorErr
    }
    limit := msg.Limit
    for limit > 0 {

>         for _, period := Terms.Period {

Reviewer 3: I'm confused by this logic please add a comment for each loop describing what's happening

Author: `CollectMsg` has a `Limit` field to define the maximum number of subscriptions to process in the handling of msg. This is to prevent the issue describe [here](https://github.com/cosmos/cosmos-sdk/issues/4642#issuecomment-513377137)

Each subscription provider maintains a separate `DueQueue` for each period they allow. The reason for this is explained [here](https://github.com/cosmos/cosmos-sdk/pull/4780#discussion_r307527230)

Added comments to loop. Let me know if they aren't sufficient

                if err != nil {
                    DeleteSubscription(subscription.Subscriber, subscription.Name)
                    limit--
                    continue
                } else {
                    // update subscription after successful payment
                    subscription.LastPaid = ctx.BlockTime
                    if subscription.Limit != -1 {
                        subscription.Limit--
                    }
                }
                // store updated subscription
                StoreSubscription(subscriptionID, subscription)
                // push paid subscription back to end of queue
                PushToBack(queue, subscriptionID)
            } else {
                // processed all due subscriptions
                // emit event indicating all due subscriptions reached with time for next duedate
                return nil
            }
        }
    }
    // limit reached.
    // emit event indicating that not all due subscriptions have been collected
    return nil
}
```

## Consequences

### Positive

- Adds a useful module to core-sdk that can be used by many future applications

### Negative

> - One more module in `x/` folder

Reviewer 3: this is a larger issue than this ADR so I wouldn't worry about it

Users can define maximum limits on how many periods the subscription is valid for. If this limit elapses before a subscription renewal, the subscription is invalidated. This is to mitigate the common problem of unused subscriptions siphoning off funds from account without the user noticing. For example, a user can subscribe to a service with a period of 3 months with a limit of 4. This subscription will automatically expire after a year unless the user manually increases/removes the limit with a second `SubscribeMsg`.

Users can also manually unsubscribe by submitting an `UnsubscribeMsg`

## Status

#### Proposed

## Implementation Changes

Introduce the following messages to `x/subscription` module:

```go
// CreateSubscriptionMsg allows Service Provider to create new subscription service.
// Generate unique serviceID from hash(name|collector)
type CreateSubscriptionMsg struct {
    Name        string          // human-readable name for subscription service
    Description string          // description for what the service provides

>     Amounts     []sdk.Coins     // amounts to be collected for each subscription period

Reviewer 2: we should allow subscription providers to append new periods to the existing ones

    Periods     []time.Duration // allowed duration of subscription periods
    Collector   sdk.AccAddress  // address that will collect subscription payments
}
```

```go
// SubscribeMsg allows subscriber account to subscribe to provided service
// If subscription to the service from user already exists, this msg is treated as renewal.

> // If msg.Limit != -1, then subscription.Limit += msg.Limit. Else, subscription.Limit = -1 (unlimited)

Reviewer 2: how can we decrease the limit? Maybe just set the passed in limit as the updated limit instead of adding to the existing limit

    Collector sdk.AccAddress // address that will collect payments
    Limit     int64          // maximum number of items to process in FIFO duequeue. If Limit = -1, try to process all due subscriptions
}
```

Create a new store with the following key-values:

`Address => []SubscriptionID // List of active/inactive subscriptions owned by the users`

`Terms:{ServiceID} => SubscriptionTerms`

`DueQueue:{Collector}{ServiceID}{Period} => LinkedList<SubscriptionID> // FIFO queue of subscriptions for a given service and period. Note if a service allows for multiple periods, each period will maintain a separate queue. CONTRACT: All due subscriptions exist before all undue subscriptions. All subscriptions in DueQueue are active`

`SubscriptionID => Subscription`

```go
// SubscriptionID is a unique identifier to a subscription struct
// hash(Name+Address)
type SubscriptionID []byte
```

```go
// Subscription Terms contains information necessary for processing subscriptions
type Terms struct {
    Name        string         // short, human-readable name of the service
    Description string         // long-form description of the service

>     Amount      sdk.Coins      // amount to be collected for each subscription period

Reviewer 2: ```suggestion
    Amounts      []sdk.Coins      // amount to be collected for each subscription period
```

>     Period      time.Duration  // duration of subscription period

Reviewer 2: ```suggestion
    Periods      []time.Duration  // duration of subscription period
```

    // adds subscription to list of user subscriptions
    AppendToUserSubscriptions(subscriptionID)
    // push to back of DueQueue for this service
    PushToBackOfDueQueue(msg.Collector, msg.ServiceID, subscriptionID)
}
```

```go
// simply deletes mapping subscriptionID => subscription and removes ID from user list
// will not mutate duequeue
func HandleUnsubscribeMsg(ctx sdk.Context, msg UnsubscribeMsg) {
    DeleteSubscription(msg.Subscriber, msg.ServiceID)
}
```

```go
func HandleCollectMsg(ctx sdk.Context, msg CollectMsg) {
    if !SubscriptionExists(msg.ServiceID) {
        return NoSuchSubscriptionErr
    }
    Terms := GetTerms(ctx, msg.ServiceID)
    if msg.Collector != Terms.Collector {
        return InvalidCollectorErr
    }
    limit := msg.Limit
    // Only process a limited number of subscriptions defined in CollectMsg
    // This is to prevent msg handler from panicing with OutOfGasException if DueQueue gets too large
    for limit > 0 {
        // Each subscription period maintains its own queue
        // These queues are processed in a round robin fashion until all due subscriptions are collected
        // or msg.Limit has been reached

>         for _, period := Terms.Period {

Reviewer 2: add msg to collect from specific queue instead of round robin on all
